### PR TITLE
Throttle policy template builder and throttle extension modified

### PR DESCRIPTION
**Pull request by dinusha92**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]
Policy template builders added.
Throttle siddhi extension was updated.
ThrottleDeploymentManager removed.
-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+                context = new VelocityContext();
+                setConstantContext(context);
+                context.put("pipelineItem", pipeline);
+                context.put("policy", policy);
+
+                context.put("quotaPolicy", pipeline.getQuotaPolicy());
+                context.put("pipeline", "condition_" + pipeline.getId());
+
+                String conditionString = getPolicyCondition(pipeline.getConditions());
+                //                    conditionsSet.add(conditionString);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: why we have a commentted code?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        Template template = velocityengine.getTemplate(getTemplatePathForAPI());
+        StringWriter writer;
+        VelocityContext context;
+
+        if (policy.getPipelines() != null) {
+
+            for (Pipeline pipeline : policy.getPipelines()) {
+                context = new VelocityContext();
+                setConstantContext(context);
+                context.put("pipelineItem", pipeline);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: let's move these strings to constants

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Siddhi query builder for API throttle policy.
+ */
+public class APIThrottlePolicyTemplateBuilder extends ThrottlePolicyTemplateBuilder {
+
+    private static final Log log = LogFactory.getLog(ThrottlePolicyTemplateBuilder.class);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: correct class?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        if (log.isDebugEnabled()) {
+            log.debug("Generating policy for apiLevel :" + policy.toString());
+        }
+        Map<String, String> policyArray = new HashMap<String, String>();
+        VelocityEngine velocityengine = initVelocityEngine();
+        Template template = velocityengine.getTemplate(getTemplatePathForAPI());
+        StringWriter writer;
+        VelocityContext context;
+
+        if (policy.getPipelines() != null) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: let's add more comments

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+                context.put("quotaPolicy", pipeline.getQuotaPolicy());
+                context.put("pipeline", "condition_" + pipeline.getId());
+
+                String conditionString = getPolicyCondition(pipeline.getConditions());
+                //                    conditionsSet.add(conditionString);
+
+                context.put("condition", " AND " + conditionString);
+                writer = new StringWriter();
+                template.merge(context, writer);
+                if (log.isDebugEnabled()) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: what's the intention of this debug log?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+                    log.debug("Policy : " + writer.toString());
+                }
+
+                String policyName =
+                        PolicyConstants.POLICY_LEVEL_RESOURCE + "_" + policy.getPolicyName() + "_condition_" + pipeline
+                                .getId();
+                policyArray.put(policyName, writer.toString());
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: add more debug logs with meningful content

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    /**
+     * Generate default policy for api level throttling
+     *
+     * @param policy Policy with level 'api'. isAcrossAllUsers() method in policy is used to identify the level in
+     *               the api level. Policy can have multiple pipelines and a default condition which will be used as
+     *               else condition
+     * @return throttle policies for default api
+     * @throws APITemplateException throws if generation failure occur
+     */
+    public String getThrottlePolicyForAPILevelDefault(APIPolicy policy) throws APITemplateException {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: let's change the method name to meanngful name

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+        if (log.isDebugEnabled()) {
+            log.debug("Generating policy for apiLevel :" + policy.toString());
+        }
+        Set<String> conditionsSet = new HashSet<String>();
+
+        VelocityEngine velocityengine = initVelocityEngine();
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: add more debug logs

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+                    conditionsSet.add(conditionString);
+                }
+            }
+        }
+
+        // for default one
+        context = new VelocityContext();
+        setConstantContext(context);
+        //default policy is defined as 'elseCondition'
+        context.put("pipeline", "elseCondition"); //// constant
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: move these strings to constants

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+        Set<String> conditionsSet = new HashSet<String>();
+
+        VelocityEngine velocityengine = initVelocityEngine();
+        Template template = velocityengine.getTemplate(getTemplatePathForAPIDefaultPolicy());
+        StringWriter writer;
+        VelocityContext context;
+
+        List<Pipeline> pipelines = policy.getPipelines();
+
+        if (pipelines != null) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: add more comments

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        context.put("quotaPolicy", policy.getDefaultQuotaPolicy());
+        String conditionSetString = getConditionForDefault(conditionsSet);
+        if (!StringUtils.isEmpty(conditionSetString)) {
+            context.put("condition", " AND " + conditionSetString);
+        } else {
+            context.put("condition", "");
+        }
+        writer = new StringWriter();
+        template.merge(context, writer);
+        if (log.isDebugEnabled()) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: what's the use of the debug log?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+        }
+
+        return conditionString;
+    }
+
+    @Override
+    public Map<String, String> getThrottlePolicyTemplate() {
+        try {
+            templateMap = getThrottlePolicyForAPILevel(apiPolicy);
+            templateMap.put("default", getThrottlePolicyForAPILevelDefault(apiPolicy));
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: move to constants

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/ApplicationThrottlePolicyTemplateBuilder.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.velocity.Template;
+import org.apache.velocity.VelocityContext;
+import org.apache.velocity.app.VelocityEngine;
+import org.wso2.carbon.apimgt.core.models.policy.ApplicationPolicy;
+
+import java.io.StringWriter;
+import java.util.Map;
+/**
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/ApplicationThrottlePolicyTemplateBuilder.java]: add a new line

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/ApplicationThrottlePolicyTemplateBuilder.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+ * Siddhi query builder for application throttle policy.
+ */
+public class ApplicationThrottlePolicyTemplateBuilder extends ThrottlePolicyTemplateBuilder {
+
+    private static final Log log = LogFactory.getLog(ThrottlePolicyTemplateBuilder.class);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/ApplicationThrottlePolicyTemplateBuilder.java]: correct class

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/ApplicationThrottlePolicyTemplateBuilder.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    /**
+     * Generate application level policy.
+     *
+     * @param policy policy with level 'app'. Multiple pipelines are not allowed. Can define more than one condition
+     *               as set of conditions. all these conditions should be passed as a single pipeline
+     * @return throttle policies for app level
+     * @throws APITemplateException throws if generation failure occur
+     */
+    public String getThrottlePolicyForAppLevel(ApplicationPolicy policy) throws APITemplateException {
+        StringWriter writer = new StringWriter();
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/ApplicationThrottlePolicyTemplateBuilder.java]: add more comments debug logs and move strings to constanjts

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+import java.io.StringWriter;
+import java.util.Map;
+
+/**
+ * Siddhi query template builder for Custom throttle policy.
+ */
+public class CustomThrottlePolicyTemplateBuilder extends ThrottlePolicyTemplateBuilder {
+    private static final String POLICY_VELOCITY_GLOBAL = "throttle_policy_template_global";
+    private static final Log log = LogFactory.getLog(ThrottlePolicyTemplateBuilder.class);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java]: correct class

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    public String getThrottlePolicyForGlobalLevel(CustomPolicy policy) throws APITemplateException {
+        StringWriter writer = new StringWriter();
+
+        if (log.isDebugEnabled()) {
+            log.debug("Generating policy for globalLevel :" + policy.toString());
+        }
+
+        try {
+            VelocityEngine velocityengine = initVelocityEngine();
+
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java]: why new lines?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+     * Generate policy for global level.
+     *
+     * @param policy policy with level 'global'. Multiple pipelines are not allowed. Can define more than one condition
+     *               as set of conditions. all these conditions should be passed as a single pipeline
+     * @return policy template as a string
+     * @throws APITemplateException throws if any error occurred
+     */
+    public String getThrottlePolicyForGlobalLevel(CustomPolicy policy) throws APITemplateException {
+        StringWriter writer = new StringWriter();
+
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java]: add  more comments and debug logs

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+            Template template = velocityengine.getTemplate(getTemplatePathForGlobal());
+
+            VelocityContext context = new VelocityContext();
+            setConstantContext(context);
+            context.put("policy", policy);
+            if (log.isDebugEnabled()) {
+                log.debug("Policy : " + writer.toString());
+            }
+            template.merge(context, writer);
+        } catch (Exception e) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java]: why we catching a generic ecception?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+            log.error("Velocity Error", e);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java]: Add descripting error messages

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+        return writer.toString();
+    }
+
+    private String getTemplatePathForGlobal() {
+        return policyTemplateLocation + POLICY_VELOCITY_GLOBAL + ".xml";
+    }
+
+    @Override public Map<String, String> getThrottlePolicyTemplate() {
+        try {
+            templateMap.put("default", getThrottlePolicyForGlobalLevel(customPolicy));
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java]: to constants

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+        } catch (APITemplateException e) {
+            String errorMessage = "Error while creating template for advaced throttle policy.";
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/CustomThrottlePolicyTemplateBuilder.java]: why we dont throw the exception upwards?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/SubscriptionThrottlePolicyTemplateBuilder.java [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+import org.wso2.carbon.apimgt.core.models.policy.SubscriptionPolicy;
+
+import java.io.StringWriter;
+import java.util.Map;
+/**
+ * Siddhi query builder for Subscription throttle policy.
+ */
+public class SubscriptionThrottlePolicyTemplateBuilder extends ThrottlePolicyTemplateBuilder {
+
+    private static final Log log = LogFactory.getLog(ThrottlePolicyTemplateBuilder.class);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/SubscriptionThrottlePolicyTemplateBuilder.java]: correct class

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/SubscriptionThrottlePolicyTemplateBuilder.java [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+     * @return throttle policies for subscription level
+     * @throws APITemplateException throws if generation failure occur
+     */
+    public String getThrottlePolicyForSubscriptionLevel(SubscriptionPolicy policy) throws APITemplateException {
+        StringWriter writer = new StringWriter();
+
+        if (log.isDebugEnabled()) {
+            log.debug("Generating policy for subscriptionLevel :" + policy.toString());
+        }
+
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/SubscriptionThrottlePolicyTemplateBuilder.java]: empty lines?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/SubscriptionThrottlePolicyTemplateBuilder.java [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+        VelocityEngine velocityengine = initVelocityEngine();
+        Template t = velocityengine.getTemplate(getTemplatePathForSubscription());
+
+        VelocityContext context = new VelocityContext();
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/SubscriptionThrottlePolicyTemplateBuilder.java]: move these to contants and add more debug logs comments

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java (hunk 5 above)
Comment (dinusha92) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: This is to get the debug log for generated siddhi app for the respective policy

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java (hunk 11 above)
Comment (dinusha92) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIThrottlePolicyTemplateBuilder.java]: This is to get the debug log for generated siddhi app for the respective policy

**Comment by harsha89** [comment]
Looks good. Please add test cases as well.

**Pull request merged**